service/dap/server.go

```diff
@@ -0,0 +1,420 @@
+	"github.com/go-delve/delve/service"
+	"github.com/go-delve/delve/service/api"
+	"github.com/go-delve/delve/service/debugger"
+	"github.com/google/go-dap"
+	"github.com/sirupsen/logrus"
+)
+
+// Server implements a DAP server that can accept a single client for a single
+// debug session. It does not support restarting.
+type Server struct {
+	// config is all the information necessary to start the debugger and server.
+	config *service.Config
+	// listener is used to serve HTTP.
+	listener net.Listener
+	// conn is accepted client connection.
+	conn net.Conn
+	// reader is used to read requests.
+	reader *bufio.Reader
+	// debugger is the debugger service.
+	debugger *debugger.Debugger
+	// log is used for structured logging.
+	log *logrus.Entry
+	// stopOnEntry is set to automatically stop program after start
+	stopOnEntry bool
+}
+
+// NewServer creates a new DAP Server.
+func NewServer(config *service.Config) *Server {
+	logger := logflags.DAPLogger()
+	logflags.WriteDAPListeningMessage(config.Listener.Addr().String())
+	return &Server{
+		config:   config,
+		listener: config.Listener,
+		log:      logger,
+	}
+}
+
+// Stop stops the DAP debugger service.
+// It kills the target process if it was launched by the debugger.
+func (s *Server) Stop() error {
+	s.listener.Close()
+	if s.debugger != nil {
+		kill := s.config.AttachPid == 0
+		return s.debugger.Detach(kill)
+	}
+	return nil
+}
+
+// Run starts a debugger and exposes it with an HTTP server. The debugger
+// itself can be stopped with the `detach` API. Run blocks until the HTTP
+// server stops.
+func (s *Server) Run() {
+	go func() {
+		// For now, do not support multiple clients in dap mode, serially or in parallel.
+		// The server should be restarted for every new debug session.
+		// TODO(polina): allow new client connections for new debug sessions?
+		defer s.listener.Close()
+		conn, err := s.listener.Accept()
+		if err != nil {
+			panic(err)
+		}
+		s.conn = conn
+		go s.serveDAPCodec()
+	}()
+}
+
+func (s *Server) serveDAPCodec() {
+	defer func() {
+		if s.config.DisconnectChan != nil {
+			close(s.config.DisconnectChan)
+		}
+		s.conn.Close()
+	}()
+	s.reader = bufio.NewReader(s.conn)
+	for {
+		request, err := dap.ReadProtocolMessage(s.reader)
+		// TODO(polina): Differentiate between connection and decoding
+		// errors. If we get an unfamiliar request, should we respond
+		// with a no-op or an ErrorResponse?
+		if err != nil {
+			if err != io.EOF {
+				s.log.Error("DAP error:", err)
+			}
+			break
+		}
+		s.handleRequest(request)
+	}
+}
+
+func (s *Server) handleRequest(request dap.Message) {
+	jsonmsg, _ := json.Marshal(request)
+	s.log.Debug("[<- from client]", string(jsonmsg))
+
+	switch request := request.(type) {
+	case *dap.InitializeRequest:
+		s.onInitializeRequest(request)
+	case *dap.LaunchRequest:
+		s.onLaunchRequest(request)
+	case *dap.AttachRequest:
+		s.sendUnsupportedErrorResponse(request.Request)
+		//s.onAttachRequest(request)
```

Comment:
don't leave commented-out code in, remove it for now (can always copy it from git history)